File: app-source/set_finder_icon.m
#import <Cocoa/Cocoa.h>

int main(int argc, const char *argv[]) {
  if (argc != 3) return 64;
  @autoreleasepool {
    NSString *imagePath = [NSString stringWithUTF8String:argv[1]];
    NSString *targetPath = [NSString stringWithUTF8String:argv[2]];
    NSImage *image = [[NSImage alloc] initWithContentsOfFile:imagePath];
    if (image == nil) return 65;
    BOOL succeeded = [[NSWorkspace sharedWorkspace] setIcon:image forFile:targetPath options:0];
    return succeeded ? 0 : 1;
  }
}

File: extension/sketchup_icon_keeper/set_finder_icon.js
/*
 * Apply a PNG as a custom Finder icon.
 *
 * This script is intentionally distributed as readable source. macOS runs it
 * through its built-in JavaScript for Automation interpreter.
 */
ObjC.import('AppKit');

function run(argv) {
  if (argv.length !== 2) {
    throw new Error('Expected a preview image and a target file.');
  }

  const imagePath = $(argv[0]).stringByStandardizingPath;
  const targetPath = $(argv[1]).stringByStandardizingPath;
  const image = $.NSImage.alloc.initWithContentsOfFile(imagePath);

  if (!image) {
    throw new Error('The embedded preview is not a valid image.');
  }

  const succeeded = $.NSWorkspace.sharedWorkspace
    .setIconForFileOptions(image, targetPath, 0);

  if (!succeeded) {
    throw new Error('macOS could not set the Finder icon.');
  }
}
